module/zfs/zfs_vnops.c

```diff
@@ -3367,10 +3377,8 @@ zfs_setattr(vnode_t *vp, vattr_t *vap, int flags, cred_t *cr,
 
 	if (mask & AT_SIZE) {
 		err = zfs_zaccess(zp, ACE_WRITE_DATA, 0, skipaclchk, cr);
-		if (err) {
-			ZFS_EXIT(zfsvfs);
-			return ((EROFS));
-		}
+		if (err)
+			goto out3;
```

Comment:
Is omitting the ZFS_EXIT(zfsvfs) here deliberate?    Also, this propagates whatever err is rather than setting it to EROFS.

Reply:
Oh, I see the ZFS_EXIT after out3: now.  Sorry, was blind.

Reply:
This is modeled after ZOL's version, where they decided sending `err` back from whatever function.. You already figured out ZFS_EXIT